Voicemeeter x.1.2.x

pblivingston opened the pull request:
## What

Stages upcoming changes to Voicemeeter. EQ and VBAN affected.

### Adds

Bus EQ channel `trim` and `delay`
```powershell
$vmr.bus[i].eq.channel[j].trim = [float]
$vmr.bus[i].eq.channel[j].delay = [float]
```

VBAN video outstream
```powershell
$vmr.vban.outstream[i].vfps = [int]
$vmr.vban.outstream[i].vformat = [string]   # 'png', 'jpg'
$vmr.vban.outstream[i].vquality = [int]
$vmr.vban.outstream[i].vcursor = [bool]
$vmr.vban.outstream[i].route = [int]
```

VBAN MIDI outstream `route`
```powershell
$vmr.vban.outstream[i].route = [string]
# 'none' = 0, 'midi_in' = 1, 'aux_in' = 2, 'vban_in' = 4, 'all_in' = 7, 'midi_out' = 8
```

### Changes

- `on`, `name`, `ip` now read/write on all streams
- Simple ranges of consecutive integers moved to `AddIntMembers`
Retained warning for `sr` and anything with atypical behavior, consistent with the rest of the module
Retained warning for `port` because Voicemeeter will allow this to be set below 1024
- Audio instream/outstream divergent behavior via 'if' dropped in favor of explicit separation in derived classes

## Before ready for review

- [x] Voicemeeter x.1.2.x release
- [x] Pester tests pass
- [x] CHANGELOG updated

pblivingston committed:
eq x.1.2.0

changes staged
- bus channel trim
- bus channel delay
- cell gain range

pblivingston committed:
vban x.1.2.0

changes staged
- on, name, ip partial write-only resolved
- simple ranges of consecutive integers moved to AddIntMembers
- audio instream/outstream divergent behavior separated
- midi outstream route, string
- new video outstream

pblivingston committed:
video, midi outstream routes

- VMR bug: video outstream route is write-only; test commented
- midi outstream route test was improperly int; moved to string

prelim tests for potato pass

pblivingston committed:
recorder tests, tests pass

more reliable way to locate the recording

pester tests pass for all kinds
manual tests pass for all kinds
- video vban.outstream.route
- eq.channel.cell.gain range

pblivingston committed:
Update CHANGELOG.md

updated with x.1.2.x changes

pblivingston commented:
`$vmr.vban.outstream[9].route` is write-only per a VMR bug. Pester test is commented for future use.

`Eq.Channel.Cell.Gain` range updated in README.

PR is ready for review.

pblivingston commented:
I just noticed the GUI only allows selecting specific values for `vfps`: `1, 2, 3, 5, 10, 20, 24, 25, 30`, although VMR does allow setting this with more granularity and the GUI will reflect this. I haven't tested an actual video stream, but if options outside that set do function, it should be fine to leave as-is. Let me know if you'd prefer an out-of-range warning for values outside the set presented in the GUI.

pblivingston commented:
I tested with an actual video stream - `vfps` values not shown in the GUI function properly. Video stream `FPS` in VBAN Settings in the receiving VBAN-Screen application reflects the expected value and the stream displays properly.

onyx-and-iris commented on lib/io.ps1:
This will technically add trim,delay members to strip eq channels as well. If we want to be more precise we could instead pass EqChannel a reference to the parent EQ object and use $parent.kindOfEq to add trim,delay to bus channels only. What do you think?

If we did this then $this.eqId would become $this.parent.identifier() I guess.

pblivingston replied:
How does the recent change look for this?

onyx-and-iris replied:
that looks fine to me.

onyx-and-iris replied:
I was just taking another look at this snippet, what do you think about passing references to parent objects to both EqChannel and EqCell, like this:

```powershell
class EqChannel : IRemote {
    [System.Collections.ArrayList]$cell
    [Object]$eq
    
    EqChannel ([int]$index, [Object]$eq) : base ($index, $eq.remote) {
        $this.eq = $eq

        if ($eq.kindOfEq -eq 'Bus') { AddFloatMembers -PARAMS @('trim', 'delay') }

        $this.cell = @()
        $cellCount = $this.remote.kind.cells
        for ($c = 0; $c -lt $cellCount; $c++) {
            $this.cell.Add([EqCell]::new($c, $this))
        }
    }

    [string] identifier () {
        return '{0}.Channel[{1}]' -f $this.eq.identifier(), $this.index
    }
}

class EqCell : IRemote {
    [Object]$channel
    
    EqCell ([int]$index, [Object]$channel) : base ($index, $channel.remote) {
        $this.channel = $channel

        AddBoolMembers -PARAMS @('on')
        AddIntMembers -PARAMS @('type')
        AddFloatMembers -PARAMS @('f', 'gain', 'q')
    }

    [string] identifier () {
        return '{0}.Cell[{1}]' -f $this.channel.identifier(), $this.index
    }
}
```

This way we're composing the objects similarly.

pblivingston replied:
That makes sense. If we ever need `kindOfEq` in `EqCell` in the future, it gives us access through `$this.channel.eq.kindOfEq`, which is intuitive and convenient.

onyx-and-iris replied:
alright great, I'll push that later.

onyx-and-iris commented:
> Simple ranges of consecutive integers moved to AddIntMembers

So the proposal here is to drop the warnings for quality,route members for VbanAudio streams? Any other properties it affects? Just so I'm clear, is the idea here to clean up the class/improve readability?

onyx-and-iris commented:
> Audio instream/outstream divergent behavior via 'if' dropped in favor of explicit separation in derived classes

This is a good move, concrete implementations should be moved into concrete classes. Approved.

onyx-and-iris commented:
> I tested with an actual video stream - `vfps` values not shown in the GUI function properly. Video stream `FPS` in VBAN Settings in the receiving VBAN-Screen application reflects the expected value and the stream displays properly.

In this case I think we should not add any bounds warnings. We could revisit this if it makes sense once the Remote API documentation has been updated.

pblivingston commented:
> > Simple ranges of consecutive integers moved to AddIntMembers
> 
> So the proposal here is to drop the warnings for quality,route members for VbanAudio streams? Any other properties it affects? Just so I'm clear, is the idea here to clean up the class/improve readability?

Yes, just those members and the goal is improved readability. It's consistent with the rest of the module as well, which I believe contributes to readability. If you'd prefer to keep the warnings for them, it might be a good idea to review the rest of the module for properties that also might be important enough to have a warning.

I've been considering adding kind checks for properties, child objects, etc. throughout the module - if you'd like to consider warnings for certain properties, this would be a good opportunity to discuss those. I personally think `port` is the only property like this that warrants a warning, but let me know your thoughts.

pblivingston committed:
bus trim, delay

- now pass Eq object reference to Channel
- get remote from Eq
- added kindOfEq check for trim and delay

for safety, float pester tests for potato pass

onyx-and-iris commented:
I'm going to speak a truth about the warnings which may or may not surprise you. I really only added them for the convenience of the user, however, the README (or if it didn't exist the official Remote API documentation) is the contract of usage, meaning if a user were to attempt to use a property/function outside of what is explicitly stated in the README then they're violating the contract. In other words, the warnings could just as well not be there, I only added them as a nicety. Having said that, they are brittle and like comments in code prone to go out of date (if the API were to change and we weren't to notice it), so they come with a downside too.

I don't think we should worry too much about warnings personally.

pblivingston commented:
That's consistent with my intuition and I agree.

onyx-and-iris commented:
I'm basically happy with everything in this PR, you've included some good changes as well as additions.

Just to let you know, I won't be available again until after the turn of the year due to family commitments.

pblivingston commented:
Okay great! I hope you and yours enjoy the holidays!

onyx-and-iris commented:
Thanks, you too.
